proxy/proxy.go

```diff
@@ -104,296 +104,330 @@ func (p *Proxy) Handle(w http.ResponseWriter, r *http.Request) {
 
-		//redirectURL is used for auth service as a target of successful auth redirect
-		redirectURL, err := url.ParseRequestURI(fmt.Sprintf("%s%s", strings.TrimRight(p.redirect, "/"), r.URL.Path))
+		log.WithField("ns", ns).Info("Loaded OSO token")
+
+		osoToken, err := GetOSOToken(p.authURL, pci.ClusterURL, osioToken)
 		if err != nil {
 			p.HandleError(w, err)
 			return
 		}
 
-		//If the user provides OSO token, we can directly proxy
-		if _, ok := r.Header["Authorization"]; ok { //FIXME Do we need this?
-			redirect = false
+		//Login to Jenkins with OSO token to get cookies
+		jenkinsURL := fmt.Sprintf("%s://%s/", pci.Scheme, pci.Route)
+		log.WithField("ns", ns).Info(fmt.Sprintf("Logging in %s", jenkinsURL))
+		nr, _ := http.NewRequest("GET", jenkinsURL, nil)
+		nr.Header.Set("Authorization", fmt.Sprintf("Bearer %s", osoToken))
+		c := http.DefaultClient
+		nresp, err := c.Do(nr)
+		if err != nil {
+			p.HandleError(w, err)
+			return
 		}
-
-		reqURL = *r.URL
-
-		if len(r.Cookies()) > 0 { //Check cookies and proxy cache to find user info
-			for _, cookie := range r.Cookies() {
-				if strings.HasPrefix(cookie.Name, SessionCookie) {
-					if cacheVal, ok := p.ProxyCache.Get(cookie.Value); ok { //We found a session cookie in cache
-						pci := cacheVal.(ProxyCacheItem)
-						r.Host = pci.Route
-						r.URL.Host = pci.Route
-						r.URL.Scheme = pci.Scheme
-						ns = pci.NS
-						redirect = false //user is probably logged in, do not redirect
-						cacheKey = cookie.Value
-						break
-					}
-				} else if cookie.Name == CookieJenkinsIdled { //Found a cookie saying Jenkins is idled, verify and act accordingly
-					if c, ok := p.ProxyCache.Get(cookie.Value); ok {
-						pci := c.(ProxyCacheItem)
-						isIdle, err := p.idler.IsIdle(pci.NS)
-						if err != nil {
-							p.HandleError(w, err)
-							return
-						}
-						if isIdle { //If jenkins is idled, return loading page and status 202
-							w.WriteHeader(http.StatusAccepted)
-							tmplt, err := template.ParseFiles(p.indexPath)
-							if err != nil {
-								p.HandleError(w, err)
-								return
-							}
-							data := struct {
-								Message string
-								Retry   int
-							}{
-								Message: "Jenkins has been idled. It is starting now, please wait...",
-								Retry:   10,
-							}
-							log.Info("Templating index.html")
-							err = tmplt.Execute(w, data)
-							if err != nil {
-								p.HandleError(w, err)
-								return
-							}
-							p.RecordStatistics(pci.NS, time.Now().Unix(), 0) //FIXME - maybe do this at the beginning?
-						} else { //If Jenkins is running, remove the cookie
-							cookie.Expires = time.Unix(0, 0)
-							http.SetCookie(w, cookie)
-						}
-						return //Do not proceed furter - everything is written to ResponseWriter
-					}
+		if nresp.StatusCode == http.StatusOK {
+			for _, cookie := range nresp.Cookies() {
+				if cookie.Name == CookieJenkinsIdled {
+					continue
 				}
+				http.SetCookie(w, cookie)
```

Comment:
So just to sort this out in my mind, the cookie is just an optimisation which allows for short circuit JWT decoding, tenant lookup, etc. In theory, things would work without the cookie, correct?

Reply:
The flow is 

1. redirect to auth to get JWT token
2. Get user info and OSO token
3. Login to Jenkins with OSO token
4. Find the newly generated JSESSION cookie and pass it back to user - Jenkins uses it to keep the user's session and not to force him re-login for every request, afaiu - which is why we need to tak the cookie from Proxy's login request to Jenkins and pass it back to user so that it gets stored in the browser - as if the user himself logged into Jenkins